code/__HELPERS/mobs.dm

```diff
@@ -329,6 +329,96 @@ GLOBAL_LIST_EMPTY(species_list)
 	if(interaction_key)
 		LAZYREMOVE(user.do_afters, interaction_key)
 
+/**
+ * Timed action involving one mob user. Target is optional. The amount of time left can be changed dynamically with a callback
+ *
+ * Checks that `user` does not move, change hands, get stunned, etc. for the
+ * given `delay`. Returns `TRUE` on success or `FALSE` on failure.
+ * Interaction_key is the assoc key under which the do_after is capped, with max_interact_count being the cap. Interaction key will default to target if not set.
+ * dynamic_timer_change returns when invoked either nothing if it doesnt need to change the length of the timer or the new delay if it does, which we then use to change the endtime and adjust it for how much progress the timer made
+ */
+/proc/do_after_dynamic(mob/user, delay, atom/target, timed_action_flags = NONE, progress = TRUE, datum/callback/extra_checks, interaction_key, max_interact_count = 1, datum/callback/dynamic_timer_change)
```

Comment:
Why is this what looks to just be another slightly edited do_after copy/paste? Surely there's a better way.

Reply:
i asked if there was a way to do what i wanted with what we have and was told that there isnt a way to synchronize the time in a do_after depending on how many people attempt it on the same object. should i merge this into another do_after?

Reply:
I meant that we have tons and tons of just C+P of do_after, it's hard to make a change in one because you have to do it everywhere.

I think all the contents should be split up into more granular procs that they all call instead.

Reply:
did you want me to do that in this pr?

Reply:
I think, at the very least, do it to `do_after` and this in this PR.